src/utils/parser/getScopes.js

```diff
@@ -0,0 +1,281 @@
+  bindings: any
+};
+
+export type ScopeDeclarations = ScopeDeclaration[];
+
+type GeneratedScope = {
+  start: Location,
+  end: Location,
+  type: string,
+  bindings: any,
+  children: ?(GeneratedScope[])
+};
+
+type TempScope = {
+  type: string,
+  parent: TempScope | null,
+  scopes: Array<TempScope>,
+  loc: BabelLocation,
+  names: any
+};
+
+function _createTempScope(
+  type: string,
+  parent: TempScope | null,
+  loc: BabelLocation
+): TempScope {
+  let result = {
+    type,
+    parent,
+    scopes: [],
+    loc: loc,
+    names: Object.create(null)
+  };
+  if (parent) {
+    parent.scopes.push(result);
+  }
+  return result;
+}
+
+function isNode(node?: Node, type: string): boolean {
+  return node ? node.type === type : false;
+}
+
+function _getFunctionScope(scope: TempScope): TempScope {
+  let s = scope;
+  while (s.type !== "Function" && s.type !== "Script") {
+    if (!s.parent) {
+      return s;
+    }
+    s = s.parent;
+  }
+  return s;
+}
+
+function _fromBabelLocation(
+  location: BabelLocation,
+  sourceId: SourceId
+): Location {
+  return {
+    sourceId,
+    line: location.line,
+    column: location.column
+  };
+}
+
+function _parseDeclarator(
+  declaratorId: Node,
+  targetScope: TempScope,
+  kind: string
+) {
+  if (isNode(declaratorId, "Identifier")) {
+    targetScope.names[declaratorId.name] = { kind, refs: [] };
+    return;
+  } else if (isNode(declaratorId, "ObjectPattern")) {
+    declaratorId.properties.forEach(prop => {
+      _parseDeclarator(prop.key, targetScope, kind);
+    });
+  } else if (isNode(declaratorId, "ArrayPattern")) {
+    declaratorId.elements.forEach(item => {
+      _parseDeclarator(item, targetScope, kind);
+    });
+  } else if (isNode(declaratorId, "RestElement")) {
+    _parseDeclarator(declaratorId.argument, targetScope, kind);
+  }
+}
+
+function _recordIdentifierPosition(
+  scope: TempScope,
+  name: string,
+  loc: BabelLocation
+) {
+  for (let s = scope; s; s = s.parent) {
+    if (name in s.names) {
+      scope.names[name].refs.push(loc.start);
+      break;
+    }
+  }
+}
+
+function _toGeneratedScopes(
+  children: TempScope[],
+  sourceId: SourceId
+): ?(GeneratedScope[]) {
+  if (!children || children.length === 0) {
+    return undefined;
+  }
+  return children.map(scope => {
+    let bindings = Object.create(null);
+    Object.keys(scope.names).forEach(n => {
+      switch (scope.names[n].kind) {
+        case "var":
+        case "let":
+        case "const":
+        case "param":
+          bindings[n] = scope.names[n].refs.map(location => {
+            return _fromBabelLocation(location, sourceId);
+          });
+          break;
+      }
+    });
+    return {
+      start: _fromBabelLocation(scope.loc.start, sourceId),
+      end: _fromBabelLocation(scope.loc.end, sourceId),
+      type: scope.type.toLowerCase(),
+      bindings,
+      children: _toGeneratedScopes(scope.scopes, sourceId)
+    };
+  });
+}
+
+function _parseJSScopes(source: Source) {
```

Comment:
whey do we need to find scopes when the babel paths have a scope property?

Reply:
We need position/references of every name in particular scope. So we can match these with source maps original names.